File: 152_Maximum_Product_Subarray/152_1.py
from typing import List

class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        res = nums[0]
        max_num, min_num = nums[0], nums[0]
        for i in range(1, len(nums)):
            tmp_max, tmp_min = max_num * nums[i], min_num * nums[i]
            max_num = max(nums[i], tmp_max, tmp_min)
            min_num = min(nums[i], tmp_min, tmp_max)
            res = max(res, max_num)
        return res

File: 152_Maximum_Product_Subarray/152_1.java
class Solution {
    public int maxProduct(int[] nums) {
        int res = nums[0];
        int max = nums[0], min = nums[0];
        for (int i = 1; i < nums.length; i++) {
            int n = nums[i];
            int tmp_min = min * n;
            int tmp_max = max * n;
            max = Math.max(n, Math.max(tmp_min, tmp_max));
            min = Math.min(n, Math.min(tmp_max, tmp_min));
            res = Math.max(res, max);
        }
        return res;
    }
}
